bedrock/wordpress/models.py

```diff
@@ -58,38 +81,59 @@ def filter_by_blogs(self, *blog_slugs):
     def filter_by_tags(self, *tags):
         return self.get_queryset().filter_by_tags(*tags)
 
-    def update_posts(self, data, database='default'):
-        with transaction.atomic(using=database):
-            count = 0
-            posts = data['posts']
-            self.filter_by_blogs(data['wp_blog_slug']).delete()
-            for post in posts:
-                try:
-                    self.create(
-                        wp_id=post['id'],
-                        wp_blog_slug=data['wp_blog_slug'],
-                        date=post['date_gmt'],
-                        modified=post['modified_gmt'],
-                        title=strip_tags(post['title']['rendered']),
-                        excerpt=process_excerpt(post['excerpt']['rendered']),
-                        link=post['link'],
-                        featured_media=post['featured_media'],
-                        tags=post['tags'],
-                    )
-                    count += 1
-                except (DatabaseError, KeyError):
-                    sentry_client.captureException()
-                    raise
-
-            return count
-
-    def refresh(self, feed_id, database='default', num_posts=None):
-        data = get_posts_data(feed_id, num_posts)
-        if data:
-            return self.update_posts(data, database)
+    def update_posts(self, blog_slug, posts):
+        update_count = 0
+        del_count = 0
+        post_ids = []
+        posts_to_update = []
+        for post in posts:
+            modified = make_datetime(post['modified_gmt'])
+            post_ids.append(post['id'])
+            try:
+                obj = self.get(wp_id=post['id'], wp_blog_slug=blog_slug)
+            except BlogPost.DoesNotExist:
+                posts_to_update.append((None, post))
+            else:
+                if obj.modified != modified:
+                    posts_to_update.append((obj, post))
+
+        if not posts_to_update:
+            return 0, 0
+
+        complete_posts_data(blog_slug, posts_to_update)
+        for obj, post in posts_to_update:
+            post = post_to_dict(blog_slug, post)
+            if not post:
+                continue
+
+            try:
+                if obj:
+                    for key, value in post.iteritems():
+                        setattr(obj, key, value)
+                    obj.save()
+                else:
+                    self.create(**post)
+            except DatabaseError:
+                sentry_client.captureException()
+                raise
+
+            update_count += 1
+
+        # clean up after changes
+        if update_count:
```

Comment:
There're may be expired posts without any updates, right? In case a blog post was just deleted?

Reply:
Could be. It's not critical that these be deleted in a timely manner. This is here just to make sure the DB doesn't grow much larger than it needs to. We pull in 20 posts by default, but the most we every use on a page is 5 at the moment. The blogs are active enough that this should be okay IMHO. If you think it's worth a check every time I'm fine with doing that as well.

Reply:
I would check every time just because things change but I won't insist on that, if you feel it's ok keep is the way it is.